File: .agents/web-scripting/python-kodi/resources/lib/mbuc_kodi_security/aes_gcm.py
from __future__ import annotations

import base64
import binascii
import os
from dataclasses import dataclass
from typing import Protocol

from cryptography.exceptions import InvalidTag
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from .errors import CorporateError

PREFIX = "v1."
NONCE_SIZE = 12
TAG_SIZE = 16


class KeyProvider(Protocol):
    def get_key(self) -> bytes:
        raise NotImplementedError


@dataclass(frozen=True, slots=True)
class EnvironmentKeyProvider:
    variable_name: str = "MBUC_AES_KEY_BASE64"

    def get_key(self) -> bytes:
        encoded = os.environ.get(self.variable_name)
        if not encoded:
            raise CorporateError("missing_key", f"Environment variable {self.variable_name} is required.")
        try:
            key = base64.b64decode(encoded, validate=True)
        except (binascii.Error, ValueError) as error:
            raise CorporateError("invalid_key", f"Environment variable {self.variable_name} is not valid Base64.", cause=error) from error
        _validate_key(key)
        return key


class AesGcmService:
    def __init__(self, key_provider: KeyProvider) -> None:
        self._key_provider = key_provider

    def encrypt(self, plaintext: bytes, associated_data: bytes) -> str:
        _validate_aad(associated_data)
        key = self._key_provider.get_key()
        _validate_key(key)
        nonce = os.urandom(NONCE_SIZE)
        ciphertext_and_tag = AESGCM(key).encrypt(nonce, plaintext, associated_data)
        ciphertext, tag = ciphertext_and_tag[:-TAG_SIZE], ciphertext_and_tag[-TAG_SIZE:]
        return PREFIX + base64.urlsafe_b64encode(nonce + tag + ciphertext).rstrip(b"=").decode("ascii")

    def decrypt(self, envelope: str, associated_data: bytes) -> bytes:
        _validate_aad(associated_data)
        if not isinstance(envelope, str) or not envelope.startswith(PREFIX):
            raise CorporateError("invalid_envelope", "Unsupported crypto envelope version.")
        encoded = envelope[len(PREFIX):]
        try:
            payload = base64.b64decode(encoded + "=" * (-len(encoded) % 4), altchars=b"-_", validate=True)
        except (binascii.Error, ValueError) as error:
            raise CorporateError("invalid_envelope", "Envelope payload is not valid Base64URL.", cause=error) from error
        if len(payload) < NONCE_SIZE + TAG_SIZE:
            raise CorporateError("invalid_envelope", "Envelope payload is truncated.")
        nonce = payload[:NONCE_SIZE]
        tag = payload[NONCE_SIZE:NONCE_SIZE + TAG_SIZE]
        ciphertext = payload[NONCE_SIZE + TAG_SIZE:]
        key = self._key_provider.get_key()
        _validate_key(key)
        try:
            return AESGCM(key).decrypt(nonce, ciphertext + tag, associated_data)
        except InvalidTag as error:
            raise CorporateError("authentication_failed", "Envelope authentication failed.", cause=error) from error


def _validate_key(key: bytes) -> None:
    if len(key) != 32:
        raise CorporateError("invalid_key", "AES-256-GCM requires exactly 32 key bytes.")


def _validate_aad(associated_data: bytes) -> None:
    if not associated_data:
        raise CorporateError("invalid_aad", "Associated data is required.")

File: Biblioteca_Universal_Corporativa/web-scripting/typescript/src/aes-gcm.ts
import { createCipheriv, createDecipheriv, randomBytes } from "node:crypto";
import { CorporateError } from "./errors.js";

const PREFIX = "v1.";
const NONCE_SIZE = 12;
const TAG_SIZE = 16;

export interface KeyProvider { getKey(): Buffer; }

export class EnvironmentKeyProvider implements KeyProvider {
  public getKey(): Buffer {
    const encoded = process.env.MBUC_AES_KEY_BASE64;
    if (!encoded) throw new CorporateError("missing_key", "MBUC_AES_KEY_BASE64 is required.");
    const key = Buffer.from(encoded, "base64");
    validateKey(key);
    return key;
  }
}

export class AesGcmService {
  public constructor(private readonly keyProvider: KeyProvider) {}

  public encrypt(plaintext: Uint8Array, associatedData: Uint8Array): string {
    validateAad(associatedData);
    const key = this.keyProvider.getKey();
    validateKey(key);
    const nonce = randomBytes(NONCE_SIZE);
    try {
      const cipher = createCipheriv("aes-256-gcm", key, nonce, { authTagLength: TAG_SIZE });
      cipher.setAAD(Buffer.from(associatedData));
      const ciphertext = Buffer.concat([cipher.update(plaintext), cipher.final()]);
      const tag = cipher.getAuthTag();
      return PREFIX + Buffer.concat([nonce, tag, ciphertext]).toString("base64url");
    } catch (cause) {
      throw new CorporateError("encryption_failed", "Encryption failed.", { cause });
    } finally { key.fill(0); }
  }

  public decrypt(envelope: string, associatedData: Uint8Array): Buffer {
    validateAad(associatedData);
    if (!envelope.startsWith(PREFIX)) throw new CorporateError("invalid_envelope", "Unsupported crypto envelope version.");
    const encoded = envelope.slice(PREFIX.length);
    if (!/^[A-Za-z0-9_-]+$/u.test(encoded)) throw new CorporateError("invalid_envelope", "Envelope payload is not valid Base64URL.");
    const payload = Buffer.from(encoded, "base64url");
    if (payload.length < NONCE_SIZE + TAG_SIZE) throw new CorporateError("invalid_envelope", "Envelope payload is truncated.");
    const nonce = payload.subarray(0, NONCE_SIZE);
    const tag = payload.subarray(NONCE_SIZE, NONCE_SIZE + TAG_SIZE);
    const ciphertext = payload.subarray(NONCE_SIZE + TAG_SIZE);
    const key = this.keyProvider.getKey();
    validateKey(key);
    try {
      const decipher = createDecipheriv("aes-256-gcm", key, nonce, { authTagLength: TAG_SIZE });
      decipher.setAAD(Buffer.from(associatedData));
      decipher.setAuthTag(tag);
      return Buffer.concat([decipher.update(ciphertext), decipher.final()]);
    } catch (cause) {
      throw new CorporateError("authentication_failed", "Envelope authentication failed.", { cause });
    } finally { key.fill(0); }
  }
}

function validateKey(key: Uint8Array): void { if (key.length !== 32) throw new CorporateError("invalid_key", "AES-256-GCM requires exactly 32 key bytes."); }
function validateAad(aad: Uint8Array): void { if (aad.length === 0) throw new CorporateError("invalid_aad", "Associated data is required."); }
